File: DSA/zawar/singlyLinkedList.cpp
#include <iostream>
using namespace std;
/* run this program using the console pauser or add your own getch, system("pause") or input loop */

//Node contains two elements:
// Data
// address of next node

class Node{
	public:
	int data;
	Node* next;
	
	Node(int val){
		data = val;
		next = NULL;
	}	
};

/* Insertion: 
1. Insert at the start of the list
2. Insert at the End of the list
3. Insert at the given index of the list
*/

//Insertion at the start (O(1))
void insertAtHead(Node* &head, int data){
	Node* newNode = new Node(data);
	newNode -> next = head;
	head = newNode;
}

//Insertion at the End (O(n))
void insertATTail(Node* &head, int data){
	Node* newNode = new Node(data);
	Node* temp = head;
	while(temp -> next != NULL){
		temp = temp->next;
	}
	temp -> next = newNode;
}

//Insert at the index (O(n))
void insertAtIndex(Node* &head, int data, int place){
	Node* newNode = new Node(data);
	Node* prevNode = head;
	int i = 0;
	while(i != place-1){
		prevNode = prevNode -> next;
		i++;
	}
	newNode -> next = prevNode -> next -> next;
	prevNode -> next = newNode;
}

//Insert at the value (O(n))
void insertAtValue(Node* &head, int data_new_node, int val){
	Node* newNode = new Node(data_new_node);
	Node* temp = head;
	while(temp -> data != val){
		temp = temp -> next;	
		if(temp == NULL){
			cout << "Not found";
		}
	}
	newNode -> next = temp -> next;
	temp -> next = newNode;
}

// Delete at head
void delAtHead(Node* &head){
	Node* toDel = head;
	head = head -> next;
	delete(toDel);
}

// Delete at Tail
void delAtTail(Node* &head){
	Node* toDel = head;
	while(toDel -> next -> next != NULL){
		toDel = toDel -> next;
	}
	Node* temp = toDel -> next -> next;
	delete(temp);
	toDel->next = NULL;
}

// Displaying the List
void displayList(Node* &ptr){
	while(ptr != NULL){
		cout << ptr -> data << " -> ";
		ptr = ptr -> next;
	}
	cout << "NULL" << endl;
}


int main(int argc, char** argv) {
	
	Node* head = new Node (1);
	Node* second = new Node (3);
	Node* third =  new Node (5);
	Node* fourth = new Node (7);
	
	head -> next = second;
	second -> next = third;
	third -> next = fourth;
	
//	displayList(head);
	insertAtHead(head, 22);
	insertATTail(head, 44);
	insertAtIndex(head, 69, 3);
	insertAtValue(head, 50, 3);
	delAtHead(head);
	delAtTail(head);
	displayList(head);
	
	return 0;
}

File: DSA/abdullah/singlyLinkedlist.py
class Node:
    def __init__(self,data=None):
        self.data=data
        self.next=None

class SL:
    def __init__(self,data):
        self.head=Node(data)

    def displayList(self):
        ptr=self.head
        while(ptr is not None):
            print(ptr.data)
            ptr=ptr.next
        print("NUll")

    def insertAtbegin(self,data):
        newNode=Node(data)
        newNode.next=self.head
        self.head=newNode

    def insertAttail(self,data):
        newNode=Node(data)
        ptr=self.head
        while(ptr.next is not None):
            # print(ptr.data)
            ptr=ptr.next
        ptr.next=newNode
    def insertAtvalue(self,value,data):
        newNode=Node(data)
        ptr=self.head
        temp=self.head
        while(ptr.data!=value):
            ptr=ptr.next
            if ptr==None:
                print("Not Found")
                return 
        newNode.next=ptr.next
        ptr.next=newNode
        
    def deleleAthead(self):
        temp=self.head.next
        self.head=temp
        del(temp)
    def deleteAttail(self):
        ptr=self.head
        while(ptr.next.next):
            ptr=ptr.next
        
        del(ptr.next.next)
        ptr.next=None
        
    def updateAthead(self,data):
        self.head.data=data
    
    def updateAttail(self,data):
        ptr=self.head
        while(ptr.next):
            ptr=ptr.next
        ptr.data=data
        
        
        
            
        

    
        

li1=SL(1)
li1.insertAtbegin(2)
li1.insertAtbegin(3)
li1.insertAtbegin(4)
li1.insertAtbegin(5)
li1.insertAtbegin(7)
li1.insertAtbegin(9)
li1.insertAttail(10)

# li1.insertAtvalue(10,88)
li1.displayList()
li1.deleleAthead()
print("after")
li1.displayList()
li1.deleteAttail()
print("after")
li1.displayList()
print("after update at head")
li1.updateAthead(99)
li1.displayList()
print("after update at head")
li1.updateAttail(83)
li1.displayList()

# print("done")



